File: backend/internal/service/orders_20.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// OrderRecord20Service encapsulates business rules for OrderRecord20.
type OrderRecord20Service struct {
	repo *repository.OrderRecord20Repository
}

// NewOrderRecord20Service wires the service.
func NewOrderRecord20Service(repo *repository.OrderRecord20Repository) *OrderRecord20Service {
	return &OrderRecord20Service{repo: repo}
}

// Create validates and persists a new OrderRecord20.
func (s *OrderRecord20Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.OrderRecord20, error) {
	now := time.Now().UTC()
	rec := &models.OrderRecord20{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a OrderRecord20 by id.
func (s *OrderRecord20Service) Get(ctx context.Context, id string) (*models.OrderRecord20, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped OrderRecord20 pages.
func (s *OrderRecord20Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.OrderRecord20, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing OrderRecord20.
func (s *OrderRecord20Service) Update(ctx context.Context, actor string, rec *models.OrderRecord20) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.Touch(actor, time.Now().UTC())
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a OrderRecord20.
func (s *OrderRecord20Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds OrderRecord20 by external code prefix.
func (s *OrderRecord20Service) Search(ctx context.Context, facilityID, prefix string) ([]models.OrderRecord20, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *OrderRecord20Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}

File: frontend/src/types/orders20.ts
/** Domain types for orders / OrderRecord20 */
export type OrderRecord20Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface OrderRecord20 {
  id: string;
  externalCode: string;
  displayName: string;
  status: OrderRecord20Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface OrderRecord20CreateRequest {
  externalCode: string;
  displayName: string;
  status: OrderRecord20Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface OrderRecord20ListResponse {
  items: OrderRecord20[];
  total?: number;
}

export interface OrderRecord20StatsResponse {
  activeCount: number;
}

export function isOrderRecord20Active(rec: OrderRecord20): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatOrderRecord20Label(rec: OrderRecord20): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareOrderRecord20ByPriority(a: OrderRecord20, b: OrderRecord20): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}
